draft-ietf-mls-protocol.md

```diff
@@ -1970,19 +1963,25 @@ message at the same time, by taking the following steps:
   value in the MLSPlaintext.  Sign the MLSPlaintext using the current epoch's
   GroupContext as context.
 
-* Complete the GroupInfo by populating the following fields:
-  * Confirmed transcript hash: The confirmed transcript hash including the
-    current Commit object
-  * Interim transcript hash: The interim transcript hash including the current
-    Commit object
-  * Confirmation: The confirmation from the MLSPlaintext
+* Update the tree in the provisional state by applying the direct path
+
+* Construct a GroupInfo reflecting the new state:
+  * Group ID, epoch, tree, confirmed transcript hash, and interim transcript
+    hash from the new state
+  * The confirmation from the MLSPlaintext object
   * Sign the GroupInfo using the member's private signing key
-  * Encrypt the GroupInfo using the key and nonce derived from the `init_secret`
-    for the current epoch (see {{welcoming-new-members}})
+  * Encrypt the GroupInfo using the key and nonce derived from the `epoch_secret`
+    for the new epoch (see {{welcoming-new-members}})
 
-* For each new member in the group, compute an EncryptedKeyPackage object that
-  encapsulates the `init_secret` for the current epoch.  Construct a Welcome
-  message from the encrypted GroupInfo object and the encrypted key packages.
+* For each new member in the group:
+  * Identify the lowest common ancestor in the tree of the new member's
+    leaf node and the member sending the Commit
```

Comment:
I think there might be more math involved here. Unless it's a group of two, the direct path of the new member should be blank until it intersects with the direct path of the adder (who populated their path with the Commit). In other words, the common ancestor node has to be resolved, leading to an array of resolved nodes. The new member should be one of these resolved nodes.
To me this looks like you end up KEMing to the new member's leaf node directly.

Reply:
You're kind of making a point I should have made, namely that sending the path secret of the common ancestor as we are here is equivalent to sending the direct path -- in the direct path encryption, you would encrypt the path secret to the corresponding copath node, which would ultimately resolve to the leaf node (and possibly some others).

What this PR is doing is taking the KEMing to the new member's leaf and taking it out of the context of the DirectPath.  Where before you were doing an HPKE to the new member once for EncryptedKeyPackage and again in the DirectPath, now you just do it for the EncryptedKeyPackage.

Reply:
Sure, I think that makes sense. 

But there might be another issue: I forgot that we don't blank the direct path of new members anymore and instead use `unmerged_leaves`. This list should now be shorter and only include nodes from leaf to common ancestor. This is however something every member in the group processing the `Add` has to calculate as it is non explicit.

Reply:
> This list should now be shorter and only include nodes from leaf to common ancestor

This is already true.  Even before this PR, the DirectPath processing would reset the `unnmerged_leaves` arrays to `[]` for nodes in the direct path of the committer.

Reply:
In any case, we should be clear that you reset `unmerged_leaves` whenever you write to a node.  I don't think that needs to be blocking on this PR.